proposed/caller-unsafe.md

```diff
@@ -0,0 +1,103 @@
+
+For more precise details on the error semantics of unsafe blocks and unsafe members, the rules will mirror the rules defined for "Requires" attributes defined in [Feature attribute semantics](https://github.com/dotnet/runtime/blob/main/docs/design/tools/illink/feature-attribute-semantics.md#requiresfeatureattribute). The only addition is the presence of the `unsafe` block, which effectively provides a local `Requires` context.
+
+## Implementation
+
+In addition to compiler enforcement, the following attribute will be added for annotating unsafe members. It is an error to use this attribute directly in C#. Instead, the `unsafe` keyword should be used.
+
+```C#
+[AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Method | AttributeTargets.Property | AttributeTargets.Constructor)]
+public sealed class RequiresUnsafeAttribute : System.Attribute
+{
+}
+```
+
+### Globally-valid properties
+
+The overall goal is to ensure .NET code is "valid," in the sense that certain properties are always true. Generating a complete list of such properties is out of scope of this document. However, at least the following properties are required:
+
+* Memory safety
+* No access to uninitialized memory
+
+In this document **memory safety** is strictly defined as: code can never acquire a reference to memory that is not managed by the runtime. "Managed" here does not refer to solely to heap-allocated, garbage collected memory, but also includes stack-allocated variables that are considered allocated by the runtime, or memory that is acquired for legal use by the runtime through any other means.
+
+No access to uninitialized memory means that all memory is either never read before it has been initialized, or it has been initialized to a zero value.
+
+These properties are particularly important for security purposes. Any methods that can potentially violate these guarantees should be unsafe. Additionally, all such methods should have documentation that describes what conditions need to be satisfied to ensure that these guarentees are preserved.
+
+These properties are guaranteed by "safe" code through a combination of compiler and .NET runtime enforcement. For `unsafe` code, these properties must be guaranteed by the programmer.
+
+`unsafe` members are used to identify the places that cannot be automatically checked by the compiler and runtime for validity. Inside unsafe blocks, the programmer is responsible for ensuring that all requirements of the unsafe code are met, and that all code outside the block will have validity properly enforced by the system.
+
+
+### Examples and APIs
+
+**ArrayPool.Rent**
+
+This method is unsafe because it returns an array with unintialized memory. Code must not read the contents of the returned array without initialization.
```

Comment:
How is uninitialized defined here? Is this because the initial array is allocated with GC.AllocateUninitializedArray, or is this because someone can rent an array, write data into it, return it, and then someone ride renting gets that data?

If it's the former, I'd rather just change the Rent implementation to use new[] instead of AllocateUninitializedArray. If it's the latter, that won't help.

Reply:
The former -- uninitialized means uninitialized by the "application" vs. the OS. If the pool used regularly .NET array allocation it would be zero-init and considered initialized.

Reply:
Uninitialized memory is not the only reason why `ArrayPool.Rent/Return` should be marked as unsafe since (the lifetime of) the memory returned by `ArrayPool.Rent/Return` is not managed by the runtime. Is this your interpretation of "managed by the runtime"?

Reply:
As currently designed, we don’t have a way to properly categorize safe lifetime management. So, for the purposes of this doc, all of the lifetime issues are considered safe. Unfortunately that does leave a large category of safe constructs that are known to have security impact, but I’ll leave that in the same category as directory traversal right now, I.e. out of support of the current proposal.